Agency Settings Scaffolding

ivan-kishko opened the pull request:
## Description of the change

> Added agencies settings page
   Basic info with ability to update description
   Agency team (currently on mock)
   Supervision agency systems with ability to remove or add hardcoded ones
   Jurisdiction (just added title and description since it is ignored for now)
   Also fixed bug when you click on Metric Configuration setting menu item after initial load  

## Type of change

Type: Feature

> All pull requests must have at least one of the following labels applied (otherwise the PR will fail):

| Label                       	| Description                                                                                               	|
|-----------------------------	|-----------------------------------------------------------------------------------------------------------	|
| Type: Bug                   	| non-breaking change that fixes an issue                                                                   	|
| Type: Feature               	| non-breaking change that adds functionality                                                               	|
| Type: Breaking Change       	| fix or feature that would cause existing functionality to not work as expected                            	|
| Type: Non-breaking refactor 	| change addresses some tech debt item or prepares for a later change, but does not change functionality    	|
| Type: Configuration Change  	| adjusts configuration to achieve some end related to functionality, development, performance, or security 	|
| Type: Dependency Upgrade      | upgrades a project dependency - these changes are not included in release notes                             |

## Related issues

Closes #166 

## Checklists

### Development

**This box MUST be checked by the submitter prior to merging**:
- [x] **Double- and triple-checked that there is no Personally Identifiable Information (PII) being mistakenly added in this pull request**

These boxes should be checked by the submitter prior to merging:
- [ ] Tests have been written to cover the code changed/added as part of this pull request

### Code review

These boxes should be checked by reviewers prior to merging:

- [ ] This pull request has a descriptive title and information useful to a reviewer
- [ ] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [ ] Potential security implications or infrastructural changes have been considered, if relevant

lilidworkin commented:
@ivan-kishko this looks fantastic! I tested it out and I only have two small requests:

1. Since this page is still under development, can we "gate" it so that it only shows up in the navigation bar for folks with the RECIDIVIZ_ADMIN permission? You can check that via `userStore.permissions.includes(Permission.RECIDIVIZ_ADMIN)` (you can search our codebase for other examples of that.)

2. I noticed that when checking/unchecking the supervision systems, it was a little bit laggy. I think we're waiting to update the UI until we get a response from the backend, is that right? My preference would actually be to update the UI right away. I think that's how the app works in other places (cc @terryttsai , @mxosman who could confirm this.)

terryttsai commented on publisher/src/components/Settings/hooks.ts:
```diff
@@ -42,3 +34,21 @@ export const useSettingsSearchParams = (): [
+      const { scrollHeight } = textAreaRef;
+
+      // eslint-disable-next-line no-param-reassign
+      textAreaRef.style.height = `${scrollHeight}px`;
```
Love this useful hook!

ivan-kishko replied:
Had to remove it since it did not take initial height if message retrieved from api was big [323f5cd](https://github.com/Recidiviz/justice-counts/pull/215/commits/323f5cd38ca6818ac6927ca124c7d546c3fc7c25)

terryttsai commented on publisher/src/components/Settings/AgencySettings.tsx:
```diff
@@ -0,0 +1,226 @@
+
+  const debouncedSave = useRef(debounce(saveAgencySettings, 1500)).current;
+
+  const wordsCount = agencyDescription
```
nit: with just empty spaces as the input, the `wordsCount` is 1 instead of 0

ivan-kishko replied:
change limit to 750 characters with textarea maxLength attribute

terryttsai commented on publisher/src/components/Settings/AgencySettings.tsx:
```diff
@@ -0,0 +1,226 @@
+              // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+              debouncedSave(settings, agencyId!);
+            }}
+            onKeyPress={(e) => {
```
So this is intended to prevent users from typing beyond 150 words right? I found that I can get around this limit by pasting text. 

It also feels weird that as I start typing the 150th word, as soon as I type one character I'm not allowed to type any more characters. I would expect that I can't type anymore after I finish my 150th word and try to insert a space. 

Side note, but maybe to limit text length we should use character count instead of word count for ease and consistency with most text length restrictions?

ivan-kishko replied:
Hello Terry! Well use chars instead of words to limit user input would be great since like this we can use native text area api without all those functions that should include all edge case scenarios, but this word (instead of chars) limit was written in task description. We can ask about this @lilidworkin and @hobuobi I guess, what they think about the idea of limit input with characters.

lilidworkin replied:
I think it's fine to limit input with characters! I think 1 word ~= 5 characters, so we can limit it to 150 * 5 = 750 characters? If that makes things easier, I'm fine with that!

If they paste test that is too long, can we update the "{wordsCount}/150 words" message to say something like "{wordsCount}/150 words (only the first 150 words will be saved)" (is that true? are we truncating before saving?)

terryttsai replied:
awesome!

terryttsai commented on publisher/src/stores/AgencyStore.ts:
```diff
@@ -0,0 +1,134 @@
+
+      const description = (await response.json()) as AgencySettings;
+      runInAction(() => {
+        this.agencyDescription = description.settings[0].value;
```
Right now it seems like settings only has one object in the array, and it has the `setting_type` of `PURPOSE_AND_FUNCTIONS`, but what happens if the backend is updated and returns an array with an array length of greater than 1? What if the order is not guaranteed?

Should we key by `setting_type` instead and get the setting with the `setting_type` of `PURPOSE_AND_FUNCTIONS`?

ivan-kishko replied:
[b2ccde2](https://github.com/Recidiviz/justice-counts/pull/215/commits/b2ccde259d257f0be7158cebe0347370581a74bd) fixed

ivan-kishko replied:
and this - [a50180d](https://github.com/Recidiviz/justice-counts/pull/215/commits/a50180d5927611197a857df70af8c0d258a53119), renamed SettingType to AgencySettingType

lilidworkin replied:
Agreed with this, thanks for making this change!

ivan-kishko replied:
> @ivan-kishko I'm seeing a scroll bar on the page on my computer that looks a little weird, not sure if this is easy to fix? <img alt="Screen Shot 2022-12-13 at 10 25 41 AM" width="685" src="https://user-images.githubusercontent.com/99899591/207377473-36cd6102-a4a1-4b37-8cc2-cae69ca6d1d6.png">

[323f5cd](https://github.com/Recidiviz/justice-counts/pull/215/commits/323f5cd38ca6818ac6927ca124c7d546c3fc7c25), try now please, removed overflow-y from inner container. 

Also changed a bit userStore, since auth0userID was never initialized in constructor, made getter

lilidworkin commented on common/types.ts:
```diff
@@ -30,13 +30,16 @@ export type AgencySystems =
   | "PAROLE"
   | "PROBATION"
-  | "POST_RELEASE";
+  | "POST_RELEASE"
```
@ivan-kishko I am confirming that adding to this type constant is the right thing to do! Thank you!

lilidworkin commented on publisher/src/components/Settings/AgencySettings.tsx:
```diff
@@ -0,0 +1,226 @@
+                .join(", ")}
+            </span>
+          </AgencySettingsInfoRow>
+          {/* state_code property will be replaced with state */}
```
We can remove this comment now!

ivan-kishko replied:
[323f5cd](https://github.com/Recidiviz/justice-counts/pull/215/commits/323f5cd38ca6818ac6927ca124c7d546c3fc7c25)

lilidworkin commented on publisher/src/components/Settings/AgencySettings.tsx:
```diff
@@ -0,0 +1,226 @@
+              Supervision Populations
+            </AgencySettingsBlockTitle>
+            <AgencySettingsBlockDescription>
+              Check the supervision popluations your agency is both responsible
```
popluations -> populations

ivan-kishko replied:
[323f5cd](https://github.com/Recidiviz/justice-counts/pull/215/commits/323f5cd38ca6818ac6927ca124c7d546c3fc7c25)

lilidworkin commented on publisher/src/components/Settings/AgencySettings.tsx:
```diff
@@ -0,0 +1,226 @@
+            Systems{" "}
+            <span>
+              {currentAgencySystems
+                ?.map((system) => normalizeSystem(system))
```
If you go to the North Carolina Department of Public Safety (NCDPS) agency and select all systems, the formatting looks a bit weird: 
<img width="688" alt="Screen Shot 2022-12-13 at 10 29 02 AM" src="https://user-images.githubusercontent.com/99899591/207375040-70a0acda-84fc-43d1-9f3c-74c01f9880d7.png">

ivan-kishko replied:
[323f5cd](https://github.com/Recidiviz/justice-counts/pull/215/commits/323f5cd38ca6818ac6927ca124c7d546c3fc7c25) - fixed with align to the right

lilidworkin commented on publisher/src/components/Settings/SettingsMenu.tsx:
```diff
@@ -15,32 +15,58 @@
+  )
+    ? settingsMenuPaths
+    : settingsMenuPaths.filter(
+        (menuPath) => menuPath.path === "agency-settings"
```
Shouldn't this be reversed? `(menuPath) => menuPath.path !== "agency-settings"`

ivan-kishko replied:
it should, sorry, fixed here [323f5cd](https://github.com/Recidiviz/justice-counts/pull/215/commits/323f5cd38ca6818ac6927ca124c7d546c3fc7c25)

lilidworkin reviewed (commented):
This is looking great, thanks @ivan-kishko !

lilidworkin commented:
@ivan-kishko I'm seeing a scroll bar on the page on my computer that looks a little weird, not sure if this is easy to fix?
<img width="685" alt="Screen Shot 2022-12-13 at 10 25 41 AM" src="https://user-images.githubusercontent.com/99899591/207377473-36cd6102-a4a1-4b37-8cc2-cae69ca6d1d6.png">

lilidworkin commented:
@terryttsai I'm curious if you also see this scroll bar? @ivan-kishko  is not able to repro. I have the following set in System Preferences:

<img width="442" alt="Screen Shot 2022-12-13 at 11 46 08 AM" src="https://user-images.githubusercontent.com/99899591/207393505-353eb284-e8c5-4f4d-80ce-0cee37e9fe2a.png">

ivan-kishko commented:
> @ivan-kishko I'm seeing a scroll bar on the page on my computer that looks a little weird, not sure if this is easy to fix? <img alt="Screen Shot 2022-12-13 at 10 25 41 AM" width="685" src="https://user-images.githubusercontent.com/99899591/207377473-36cd6102-a4a1-4b37-8cc2-cae69ca6d1d6.png">

[323f5cd](https://github.com/Recidiviz/justice-counts/pull/215/commits/323f5cd38ca6818ac6927ca124c7d546c3fc7c25), try now please, removed overflow-y from inner container.

Also changed a bit userStore, since auth0userID was never initialized in constructor, made getter

terryttsai commented:
@lilidworkin  I'm not able to reproduce seeing that scrollbar. Are you using chrome?

terryttsai commented on publisher/src/stores/AgencyStore.ts:
```diff
@@ -0,0 +1,139 @@
+    this.userStore = userStore;
+    this.api = api;
+    this.currentAgency = undefined;
+    this.currentAgencySystems = undefined;
```
`currentAgencySystems` and `isAgencySupervision` can be derived from `currentAgency` as [mobx computed values](https://mobx.js.org/computeds.html). So you can replace them with something like this:

```
  get currentAgencySystems(): AgencySystems[] | undefined {
    return this.currentAgency?.systems;
  }

  get isAgencySupervision(): boolean {
    return !!this.currentAgency?.systems.find(
      (system) => system === "SUPERVISION"
    );
  }
```
and change updateAgencySettings to something like this:
```
  updateAgencySettings = (
    text: string,
    systems: AgencySystems[] | undefined
  ): AgencySettings => {
    this.settings.PURPOSE_AND_FUNCTIONS = text;
    if (this.currentAgency && systems) {
      this.currentAgency.systems = systems;
    }

    return {
      settings: [{ setting_type: "PURPOSE_AND_FUNCTIONS", value: text }],
      systems,
    };
  };
```

terryttsai replied:
That way, we only have to set `currentAgency` rather than update `currentAgency`, `currentAgencySystems`, and `isAgencySupervision`. In fact, it looks like right now `updateAgencySettings` is only updating `currentAgencySystems` but not the systems in `currentAgency` or `isAgencySupervision`

terryttsai replied:
Looks great!

lilidworkin commented on publisher/src/components/Settings/AgencySettings.tsx:
```diff
@@ -0,0 +1,232 @@
+            </a>
+            .
+          </AgencySettingsBlockDescription>
+          {agencyTeam?.map(({ name }) => (
```
Small thing, would it be possible to sort the names alphabetically?

ivan-kishko replied:
Hello @lilidworkin. Done [8eed5d6](https://github.com/Recidiviz/justice-counts/pull/215/commits/8eed5d6d416819a0caeb441f36d72b1b7cc0bb9c)

lilidworkin commented on publisher/src/stores/UserStore.ts:
```diff
@@ -149,6 +146,10 @@ class UserStore {
     return this.authStore.user?.email;
   }
 
+  get auth0UserID(): string | undefined {
```
Thanks for fixing this!

lilidworkin reviewed (approved):
Looks awesome @ivan-kishko , thank you!

terryttsai reviewed (approved):
This looks awesome!
